File: controller/test/post/plotthrusterallocation.m
clear;
close all;

index_actuation=0; % underactuated=1

path = '../data/';
Balpha = csvread(strcat(path, 'Balpha.csv'),1,0);
u = csvread(strcat(path, 'u.csv'),1,0);
alpha = csvread(strcat(path, 'alpha.csv'),1,0);
alpha_deg = csvread(strcat(path, 'alpha_deg.csv'),1,0);
dtau = csvread(strcat(path, 'tau.csv'),1,0);
rotation = csvread(strcat(path, 'rotation.csv'),1,0);


m=size(alpha,2);

figure(1); 
subplot(311);
plot(Balpha(2:end,1), '-r', 'linewidth', 2);
hold on;
plot(dtau(:,1), ':k', 'linewidth', 2);
ylabel('taux(N)');
legend('estimated force', 'desired force');
subplot(312); 
plot(Balpha(2: end,2),'-r', 'linewidth', 2);
hold on;
plot(dtau(:,2), ':k', 'linewidth', 2);
ylabel('tauy(N)') 
legend('estimated force', 'desired force') 
subplot(313);
plot(Balpha(2:end,3), '-r', 'linewidth', 2);
hold on;
plot(dtau(:,3), ':k', 'linewidth', 2);
legend('estimated force', 'desired force');
ylabel('taun(N m)')

figure(2); 
for i=1:m
    subplot(m,1,i);
    plot(u(:,i),'linewidth', 2);
    ylabel(['u ',num2str(i)]);
end

    
figure(3);
for i=1:m
    subplot(m,1,i);
    plot(alpha(:,i) *180 / pi,'linewidth', 2); hold on;
    plot(alpha_deg(:,i), ':k', 'linewidth', 2);
    legend('double angle', 'int angle');
    ylabel(['alpha(deg) ',num2str(i)]);
end

figure(4);
for i=1:m
    subplot(m,1,i);
    plot(rotation(:,i),'linewidth', 2);
    ylabel(['rotation(rpm) ',num2str(i)]);
end

File: controller/test/post/plotthrusterallocation.py
# /*
# ****************************************************************************
# * plotthrustallocation.py:
# * Illustration of results of thrust allocation
# *
# * by Hu.ZH(CrossOcean.ai)
# ****************************************************************************
# */


import pandas
import matplotlib.pyplot as plt
import math

path = '../data/'
Balpha = pandas.read_csv(path+'Balpha.csv')
alpha = pandas.read_csv(path+'alpha.csv')
u = pandas.read_csv(path+'u.csv')
alpha_deg = pandas.read_csv(path+'alpha_deg.csv')
dtau = pandas.read_csv(path+'tau.csv')
rotation = pandas.read_csv(path+'rotation.csv')

m = len(alpha.columns)


plt.figure(1, figsize=(8, 8))
plt.suptitle("Desired thrust and estimated force", fontsize=14)
plt.subplot(3, 1, 1)
plt.plot(Balpha['column 1'], '-r', lw=2)
plt.plot(dtau['column 1'], '--k', lw=2)
plt.xlabel('Sampling instant')
plt.ylabel('taux (N)')
plt.grid(True)
plt.legend(('estimated force', 'desired force'), loc='upper right')

plt.subplot(3, 1, 2)
plt.plot(Balpha['column 2'], '-r', lw=2)
plt.plot(dtau['column 2'], '--k', lw=2)
plt.xlabel('Sampling')
plt.ylabel('tauy (N)')
plt.grid(True)
plt.legend(('estimated force', 'desired force'), loc='upper right')

plt.subplot(3, 1, 3)
plt.plot(Balpha['column 3'], '-r', lw=2)
plt.plot(dtau['column 3'], '--k', lw=2)
plt.xlabel('Sampling')
plt.ylabel('taun (N*m)')
plt.grid(True)
plt.legend(('estimated force', 'desired force'), loc='upper right')

plt.figure(2, figsize=(8, 8))
plt.suptitle("thrust of each propeller", fontsize=14)
for i in range(m):
    plt.subplot(m, 1, i+1)
    plt.plot(u['column '+str(i+1)], lw=2)
    plt.xlabel('Sampling')
    plt.ylabel('u'+str(i+1))
    plt.grid(True)


plt.figure(3, figsize=(8, 8))
plt.suptitle("azimuth of each propeller", fontsize=14)
for i in range(m):
    plt.subplot(m, 1, i+1)
    plt.plot(alpha['column '+str(i+1)]*180/math.pi, '-r', lw=2)
    plt.plot(alpha_deg['column '+str(i+1)], ':k', lw=2)
    plt.legend(('double angle', 'int angle'), loc='upper right')
    plt.xlabel('Sampling')
    plt.ylabel('alpha(deg) '+str(i+1))
    plt.grid(True)


plt.figure(4, figsize=(8, 8))
plt.suptitle("rotation of each propeller", fontsize=14)
for i in range(m):
    plt.subplot(m, 1, i+1)
    plt.plot(rotation['column '+str(i+1)], lw=2)
    plt.xlabel('Sampling')
    plt.ylabel('n(rpm) '+str(i+1))
    plt.grid(True)

plt.show()
